docs/RFCS/20180411_alter_column_type.md

```diff
@@ -285,25 +286,94 @@ about why a value could not be inserted.
+    `TableDescriptor`.  Either all actions in a graph eventually
+    complete successfully, or all of them are reverted.
+  * When a graph has reached a terminal state, all of its associated
+    actions are retired from the `TableDescriptor`.
+* Randomly select an action whose dependencies have been
+  successfully completed.
+  * A `TableAction` can be marked as having been definitively completed
+    by the schema-changer code, or by invoking a
+    `TableAction.IsComplete()` function in cases where the table's
+    schema-change coordinator node has restarted.
+  * Executing actions or action graphs in parallel is currently out of
+    scope for this RFC, but could represent a significant wall-time win
+    in cases where a user is making several independent changes to a
+    table schema.
+* `Resume()` the action by:
+  * Applying the action's function to the active `TableDescriptor` to
+    produce the next `TableDescriptor` to publish when the action
+    completes.  This allows for continuous revalidation of whether
+    or not the action is still valid.
+  * Marking the `TableAction` as in-progress, persisting any updates
+    made to the `TableAction` by the `Resume()` function, and the
+    original `TableDescriptor` contents.
+  * Waiting for the success or failure of any long-running background
+    processes necessary to effect the action (e.g. index backfills).
+  * Marking the `TableAction` as being definitively completed.
+* In the event that an action graph fails, a `TableAction.Rollback()`
+   function will be invoked that is given the original and proposed
+   `TableDescriptor` values to undo the change.
+   * Simply replacing the active `TableDescriptor` with the stored,
+     original `TableDescriptor` would work, but would limit future
+     opportunities for parallel schema-change execution.
+
+With this low-level action loop in mind, any long-running or multi-step
+DDL statement that operates on a `TableDescriptor` can be expanded into
+some number of `TableAction` entries which are executed in sequence.
+The user's DDL statement can also be associated with a system job which
+reflects the execution state of the associated action graph.  This will
+allow long-running DDL statements to provide status feedback to the user
+and offer a hook for canceling their execution.
+
+Thus, an `ALTER COLUMN TYPE` command would expand to the following
+collection of actions:
+* Add the shadow column
+* Add shadow indexes
+  * These depend on the add-column `TableAction`
+* Finalize the `ALTER COLUMN TYPE`
+  * This represents a point-of-no-return for the schema change and
+    is guaranteed to operate last by depending on all previous actions.
+
+### Compatibility and upgrade path
+
+Attempting to make a mixed-version cluster operate with and without
+`TableAction`-aware nodes would be exceeding complicated, so for the
+purposes of implementing this RFC, we will maintain the existing code
+path in parallel with the proposed changes.
```

Comment:
ALTER COLUMN TYPE will be implemented only through the new TableAction code path.